File: suites/osprey-compute-kernels/reference/cases/ackermann/ackermann.ml
(* Ackermann–Péter function — deep, non-tail mutual self-recursion. *)
let add a b = a + b
let sub a b = a - b

let rec ack m n =
  if m = 0 then add n 1
  else if n = 0 then ack (sub m 1) 1
  else ack (sub m 1) (ack m (sub n 1))

let () = Printf.printf "%d\n" (ack 3 10)

File: suites/osprey-compute-kernels/reference/cases/ackermann/ackermann.rs
// Ackermann–Péter function — deep, non-tail mutual self-recursion.
fn add(a: i64, b: i64) -> i64 {
    a + b
}

fn sub(a: i64, b: i64) -> i64 {
    a - b
}

fn ack(m: i64, n: i64) -> i64 {
    if m == 0 {
        add(n, 1)
    } else if n == 0 {
        ack(sub(m, 1), 1)
    } else {
        ack(sub(m, 1), ack(m, sub(n, 1)))
    }
}

fn main() {
    println!("{}", ack(3, 10));
}
